packages config for Python UDFs (#8805)
fishtown-analytics/docs.getdbt.com: website/docs/reference/resource-configs/packages.md

---
title: packages
sidebar_label: "packages"
id: packages
---

# packages<Lifecycle status="beta" />

<VersionCallout version="1.12" /> 

<File name='functions/<filename>.yml'>

```yml
functions:
  - name: <function name>
    config:
      packages: [<string>] # optional, Python UDFs only
```

</File>

## Definition

> When creating Python UDFs, you can specify public third-party PyPI packages for your Python UDF using the optional `packages` config. In this config, you can list package names (for example, `numpy`, `pandas`) and you can pin versions (for example, `pandas==1.5.0`). The warehouse installs these packages when creating the UDF, allowing your UDF to use functionality from external Python libraries.

Reviewer: ```suggestion
When creating Python UDFs, you can use the optional `packages` config to specify public third-party PyPI packages. You can list package names (for example, `numpy` and `pandas`) and pin versions (for example, `pandas==1.5.0`). The warehouse installs these packages when it creates the UDF, so your UDF can use functionality from external Python libraries.
```

> On Snowflake, some packages are installed from the Anaconda repository, and you may need to [accept Anaconda's Terms of Service](https://docs.snowflake.com/en/developer-guide/udf/python/udf-python-packages#using-third-party-packages-from-anaconda) before using them.

Reviewer: ```suggestion
On Snowflake, some packages are installed from the Anaconda repository, and you may need to [accept Anaconda's Terms of Service](https://docs.snowflake.com/en/developer-guide/udf/python/udf-python-packages#using-third-party-packages-from-anaconda) before you can use them.
```

> Python UDFs are currently supported in Snowflake and BigQuery. Each warehouse uses a different mechanism for specifying packages. For example:

Reviewer: ```suggestion
Python UDFs are currently supported in Snowflake and BigQuery. Each warehouse uses a different mechanism to specify packages. For example:
```

<TabItem value="bigquery">

<File name='functions/schema.yml'>

```yaml
functions:
  - name: is_positive_int
    description: Returns 1 if a_string matches ^[0-9]+$, else 0
    config:
      runtime_version: "3.11"
      entry_point: main
      packages:
        - numpy
        - pandas==1.5.0
    arguments:
      - name: a_string
        data_type: string
    returns:
      data_type: integer
```

</File>

Compiled SQL:

```sql
CREATE OR REPLACE FUNCTION my_dataset.is_positive_int(a_string STRING)
  RETURNS INT64
  LANGUAGE python
  OPTIONS(
    runtime_version = "python-3.11",
    entry_point = "main",
    packages = ['numpy', 'pandas==1.5.0']
  )
AS r'''
def main(a_string):
    ...
'''
```

</TabItem>

</Tabs>

## Related documentation

- [User-defined functions](/docs/build/udfs)
- [Function properties](/reference/function-properties)
- [Function configurations](/reference/function-configs)
- [Runtime version](/reference/resource-configs/runtime-version)
- [Entry point](/reference/resource-configs/entry-point)
- [Type](/reference/resource-configs/type)
- [Volatility](/reference/resource-configs/volatility)
- [Arguments](/reference/resource-properties/function-arguments)

> - [Returns](/reference/resource-properties/returns)

Reviewer: Same as above capitalisation question

---
title: packages
sidebar_label: "packages"
id: packages
---

# packages<Lifecycle status="beta" />

<VersionCallout version="1.12" /> 

<File name='functions/<filename>.yml'>

```yml
functions:
  - name: <function name>
    config:

>       packages: [<string>] # optional, Python UDFs only

Reviewer 2: is this a string? or a list?

Author: Hey @graciegoheen, good callout! `[<string>]` is meant to indicate a list of strings, so it implicitly covers multiple values. Let me know if this should be made clearer!

Reviewer 3: List of strings is the right call! If this is how we typically denote that, then it looks good to me 🙂

```

</File>

## Definition

When creating Python UDFs, you can specify public third-party PyPI packages for your Python UDF using the optional `packages` config. In this config, you can list package names (for example, `numpy`, `pandas`) and you can pin versions (for example, `pandas==1.5.0`). The warehouse installs these packages when creating the UDF, allowing your UDF to use functionality from external Python libraries.

On Snowflake, some packages are installed from the Anaconda repository, and you may need to [accept Anaconda's Terms of Service](https://docs.snowflake.com/en/developer-guide/udf/python/udf-python-packages#using-third-party-packages-from-anaconda) before using them.

Python UDFs are currently supported in Snowflake and BigQuery. Each warehouse uses a different mechanism for specifying packages. The following table shows how they're used:

> | Warehouse  | How `packages` is used |

Reviewer 2: should we just show this in the compiled code examples instead of in a table?

Author: Added samples here: https://github.com/dbt-labs/docs.getdbt.com/pull/8805/changes/33bfd0e66ca9aac18b8cf5037b1d1526ae5cde7d
